libraries/AP_Generator/AP_Generator.h

```diff
@@ -0,0 +1,110 @@
+
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+
+#ifndef GENERATOR_ENABLED
+#define GENERATOR_ENABLED !HAL_MINIMIZE_FEATURES
+#endif
+
+#if GENERATOR_ENABLED
+
+#include <AP_Param/AP_Param.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_BattMonitor/AP_BattMonitor.h>
+
+class AP_Generator_Backend;
+
+class AP_Generator
+{
+    friend class AP_Generator_Backend;
+    friend class AP_Generator_IE_FuelCell;
+    friend class AP_Generator_RichenPower;
+
+public:
+    // Constructor
+    AP_Generator();
+
+    // Do not allow copies
+    AP_Generator(const AP_Generator &other) = delete;
+    AP_Generator &operator=(const AP_Generator&) = delete;
+
+    static AP_Generator* get_singleton();
+
+    void init(void);
+    void update(void);
+
+    // Return true if AP_Generator is enabled
+    bool enabled() const;
+
+    bool arming_checks(char *failmsg, uint8_t failmsg_len) const;
+
+    AP_BattMonitor::BatteryFailsafe update_failsafes(void);
+
+    // Enum to request reading type
+    enum class ReadingType {
+        voltage,
+        current,
+        consumed_mah,
+        fuel_remain,
+        rpm,
+    };
+
+    // Helpers to retrieve measurements
+    float get_voltage(void) const { return _voltage; }
+    float get_current(void) const { return _current; }
+    float get_fuel_remain(void) const { return _fuel_remain_pct; }
+    float get_batt_consumed(void) const { return _consumed_mah; }
+    uint16_t get_rpm(void) const { return _rpm; }
+
+    // healthy() returns true if the generator is not present, or it is
+    // present, providing telemetry and not indicating an errors.
+    bool healthy(void) { return _healthy; }
+
+    // Generator controls must return true if present in generator type
+    bool is_controllable(void);
+    bool stop(void);
+    bool idle(void);
+    bool run(void);
+
+    void send_generator_status(const GCS_MAVLINK &channel);
+
+    // Parameter block
+    static const struct AP_Param::GroupInfo var_info[];
+
+private:
+
+    // Pointer to chosen driver
+    AP_Generator_Backend* _driver_ptr;
+
+    // Parameters
+    AP_Int8 _type; // Select which generator to use
+    AP_Int32 _low_fs_bitmask;
+    AP_Int32 _crit_fs_bitmask;
+
+    enum class GenType {
+        DISABLED = 0,
+        IE_650_800 = 1,
+        IE_2400 = 2,
+        RICHENPOWER = 3,
+    };
+
+    // Helper to get param and cast to GenType
+    enum GenType get_gen_type(void) const;
```

Comment:
We don't usually add helpers for this sort of thing.

But if we are oing to have one:
```suggestion
     GenType type() const;
```